pandas/core/frame.py

```diff
@@ -217,11 +217,16 @@ def __init__(self, data=None, index=None, columns=None, dtype=None,
         if isinstance(data, DataFrame):
             data = data._data
 
+        if hasattr(data, 'to_dataframe'):  # xr.Dataset
```

Comment:
I would slightly rather skip this special check. In theory, `to_dataframe` might return something else.

I don't think it will be that much slower to use the generic dict path, though we then run into the issue that `DataFrame._init_dict` does a special check for OrderedDict when deciding whether or not to order the keys. Sadly there's no way in check whether an arbitrary Mapping type has intentionally ordered keys or not (Python-ideas discussed adding `collections.abc.Ordered` but I don't think it was implemented).